File: 26.R
# Reciprocal cycles

Args         <- commandArgs()
denominators <- ifelse(is.na(Args[6]), 999, as.numeric(Args[6]))

cycles <- sapply(seq.int(denominators), function(denominator) {
    remainder <- 1 %% denominator
    first_seen <- vector()
    digit <- 0
    while ( remainder ) {
        digit <- digit + 1
        remainder <- remainder * 10
        whole <- remainder %/% denominator
        remainder <- remainder %% denominator
        key <- paste0(whole, ':', remainder)
        if ( key %in% names(first_seen) ) {
            return(digit - first_seen[[key]])
        } else {
            first_seen[key] <- digit
        }
    }
    return(0)
})

cat(which.max(cycles), fill=TRUE)

File: 26.js
#!/usr/bin/env node

// Reciprocal cycles

var program = require('commander');

program
    .version('0.1.0')
    .description('Reciprocal cycles')
    .option('-d, --denominators <int>', 'The maximum denominator', Number, 999)
    .parse(process.argv);

var max_denominator = 0;
var max_cycle = 0;
for (var denominator = 2; denominator <= program.denominators; denominator++) {
    var remainder = 1 % denominator;
    var first_seen = {};
    var digit = 0;
    while (remainder) {
        digit++;
        remainder *= 10;
        var whole = Math.floor(remainder / denominator);
        var remainder = remainder % denominator;
        var key = whole.toString() + ':' + remainder.toString();
        if (first_seen[key]) {
            // Got cycle
            var cycle = digit - first_seen[key];
            if (cycle > max_cycle) {
                max_cycle = cycle;
                max_denominator = denominator;
            }
            remainder = 0;
        } else {
            first_seen[key] = digit;
        }
    }
}

console.log(max_denominator);
